planning/behavior_velocity_planner/src/scene_module/intersection/scene_intersection.cpp

```diff
@@ -262,11 +275,100 @@ bool IntersectionModule::checkCollision(
       continue;
     }
 
-    // ignore vehicle in ego-lane. (TODO update check algorithm)
+    // ignore vehicle in ego-lane && behind ego
     const auto object_pose = object.kinematics.initial_pose_with_covariance.pose;
     const bool is_in_ego_lane = bg::within(to_bg2d(object_pose.position), ego_poly);
     if (is_in_ego_lane) {
-      continue;  // TODO(Kenji Miyake): check direction?
+      if (!planning_utils::isAheadOf(planner_data_->current_pose.pose, object_pose)) {
+        continue;
+      }
+      // consider vehicle in ego-lane && in front of ego
+      const auto lon_vel = object.kinematics.initial_twist_with_covariance.twist.linear.x;
+      const double object_decel = planner_param_.object_expected_decel;  // NOTE: this is positive
+      const double stopping_distance = lon_vel * lon_vel / (2 * object_decel);
+
+      std::vector<geometry_msgs::msg::Point> centerpoints;
+      for (auto && p : ego_lane_with_next_lane[0].centerline())
+        centerpoints.push_back(std::move(lanelet::utils::conversion::toGeomMsgPt(p)));
+      for (auto && p : ego_lane_with_next_lane[1].centerline())
+        centerpoints.push_back(std::move(lanelet::utils::conversion::toGeomMsgPt(p)));
+      const double lat_offset =
+        std::fabs(tier4_autoware_utils::calcLateralOffset(centerpoints, object_pose.position));
+      // get the nearest centerpoint to object
+      std::vector<double> dist_obj_centerpoints;
+      for (const auto & p : centerpoints)
+        dist_obj_centerpoints.push_back(
+          tier4_autoware_utils::calcDistance2d(object_pose.position, p));
+      const int obj_closest_centerpoint_idx = std::distance(
+        dist_obj_centerpoints.begin(),
+        std::min_element(dist_obj_centerpoints.begin(), dist_obj_centerpoints.end()));
+      // find two centerpoints whose distances from `closest_centerpoint` cross stopping_distance
+      double acc_dist_prev = 0.0, acc_dist = 0.0;
+      auto p1 = centerpoints[obj_closest_centerpoint_idx];
+      auto p2 = centerpoints[obj_closest_centerpoint_idx];
+      for (unsigned i = obj_closest_centerpoint_idx; i < centerpoints.size() - 1; ++i) {
+        p1 = centerpoints[i];
+        p2 = centerpoints[i + 1];
+        acc_dist_prev = acc_dist;
+        const double delta = tier4_autoware_utils::calcDistance2d(p1, p2);
```

Comment:
Does the error accumulate because the distance is not calculated along the lane?
for example
```
const auto arc_position_p1 = lanelet::utils::getArcCoordinates(ego_lane_with_next_lane, p1);
const auto arc_position_p2 = lanelet::utils::getArcCoordinates(ego_lane_with_next_lane, p2);
const double delta = arc_position_p2.length - arc_position_p1.length;
```